File: WEEK5/bj5_1068.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');
let lineIndex = 0;

const N = parseInt(input[lineIndex++]);
const parentInfo = input[lineIndex++].split(' ').map(Number);
const deleted = parseInt(input[lineIndex++]);

const arr = Array.from({ length: N }, () => []);
let root = -1;
let leafcount = 0;

for (let now = 0; now < N; now++) {
    const parent = parentInfo[now];
    if (parent === -1) {
        root = now;
    } else {
        arr[parent].push(now);
    }
}

function DFS(number) {
    let isLeaf = true;
    for (const child of arr[number]) {
        if (child === deleted) {
            continue;
        }
        isLeaf = false;
        DFS(child);
    }
    if (isLeaf) {
        leafcount++;
    }
}

if (deleted === root) {
    console.log(0);
} else {
    DFS(root);
    console.log(leafcount);
}

File: WEEK5/bj5_1068.py
import sys

sys.setrecursionlimit(10**6)
input = sys.stdin.readline

N = int(input())
parent_info = list(map(int, input().split()))
deleted = int(input())
arr = [[] for _ in range(N)]
root = -1
leafcount = 0

for now in range(N):
    parent = parent_info[now]
    if parent == -1:
        root = now
    else:
        arr[parent].append(now)


def DFS(number):
    global leafcount
    isLeaf = True

    for child in arr[number]:
        if child == deleted:
            continue
        isLeaf = False
        DFS(child)

    if isLeaf:
        leafcount += 1


if deleted == root:
    print(0)
else:
    DFS(root)
    print(leafcount)
